### DISCUSSION TITLE
Server emitting 5 times each

### DISCUSSION THREAD

**Discussion by SpringsTea** [miguelgrinberg/Flask-SocketIO#1971] [Q&A]
I have a flask-socketio server setup through uwsgi. I had too many connections which made everything unstable, so in production it runs through uwsgi with nginx as a proxy. Here is my wsgi.ini
```
[uwsgi]
module = app:app
gevent=1000

master = true
processes = 5

socket = tmp/myproject.sock
chmod-socket = 777
vacuum = true

die-on-term = true

#logto = ./logs/wsgi/uwsgi.log
log-5xx = true
disable-logging=True

```

And my nginx config:
```
user www-data;
worker_processes auto;
pid /run/nginx.pid;
include /etc/nginx/modules-enabled/*.conf;

events {
    worker_connections 1024;
    # multi_accept on;
}

                                                           
server {		
  listen 5960;                                                  
  server_name ibnnetwork;    
  
  location / {                                                  
	  include uwsgi_params;                                     
	  uwsgi_pass unix:///var/wsgi/ibn/IBNServer/tmp/myproject.sock; 
  }                                                             
}  

http {

    ##
    # Basic Settings
    ##

    sendfile on;
    tcp_nopush on;
    types_hash_max_size 2048;
    # server_tokens off;

    # server_names_hash_bucket_size 64;
    # server_name_in_redirect off;

    include /etc/nginx/mime.types;
    default_type application/octet-stream;

    ##
    # SSL Settings
    ##

    ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
    ssl_prefer_server_ciphers on;

    include /etc/nginx/conf.d/*.conf;
    include /etc/nginx/sites-enabled/*;
}
```

The problem I'm having is that when I have my server emit, it will emit 5 times.  I'm guessing each process is emitting, and my clients are hearing all of them. GPT suggested I use clustering to solve the problem, but couldn't give a correct setup to accomplish this and I can't find _any_ documentation on the topic.

Could someone please point me in the right direction?

**Comment by miguelgrinberg** [comment]
The documentation shows working uwsgi and nginx configurations. Yours are completely different. Why is that?

I suggest you start from the documented configuration:

- [uWSGI](https://flask-socketio.readthedocs.io/en/latest/deployment.html#uwsgi-web-server)
- [nginx](https://flask-socketio.readthedocs.io/en/latest/deployment.html#using-nginx-as-a-websocket-reverse-proxy)
- [using multiple worker processes](https://flask-socketio.readthedocs.io/en/latest/deployment.html#using-multiple-workers)

**Reply by SpringsTea** [reply]
My configuration is different because that's what I've needed. I needed the additional workers because my server was getting really spotty trying to keep all of the connections, adding the workers solved this issue.
My server is strictly for websockets, removing the `uwsgi_pass` stuff for the `proxy_pass` does not work, my clients just timeout.
My understanding is that nginx needs this to understand the wsgi language but there's not a lot of info out there about this.

I've modified my nginx config to include the /socket.io location block like this:

```
location /socket.io {
        include uwsgi_params;
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        uwsgi_pass unix:///var/wsgi/ibn/IBNServer/tmp/myproject.sock;    
    }
```
Again, I had to modify it to use the uwsgi parameters otherwise my clients timeout.
Nothing has changed, my server is still emitting too many times.

Since this is an issue of my server sending data and no receiving, it doesn't really make sense that this config would be the answer anyways. What exactly should I be looking at?

**Reply by miguelgrinberg** [reply]
@SpringsTea I don't believe uwsgi_pass works with WebSocket, that is only for HTTP traffic if I remember correctly. As I said above, the configuration that I share in the documentation is verified to work. Start from that, then make the changes that you need one at a time and confirm that everything continues to work. I cannot really tell you what's wrong because you have deviated from the working configuration in several ways.

**Reply by SpringsTea** [reply]
So I've got it working with more or less the config in the documentation.
Unfortunately, my issue has not been resolved which is real weird since I didn't implement the multiple processes. I can only assume the issue is gevent related?

My ini:

```
[uwsgi]
http = :5961
gevent = 1000
master = true
http-websockets
wsgi-file = app.py
callable = app
```

And my nginx
```
server {
    listen 5960;
    server_name IBN;

    location / {
        include proxy_params;
        proxy_pass http://127.0.0.1:5961;
    }

    location /socket.io {
        include proxy_params;
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_pass http://127.0.0.1:5961/socket.io;
    }
}
```

I am monkey patching gevent so i can run a background process on a schedule. It collects data from sql and then emits a message out that it's complete to all ready clients. So I have this in my entry point file:
```
from gevent import monkey
monkey.patch_all()

import modules.server.scheduler as scheduler
```

And then in the scheduler I have a APScheduler that runs a function that emits. It creates it's socket connection like this
```
from flask_socketio import SocketIO
socketio = SocketIO(message_queue="redis://{}:{}/0".format(config.REDIS_SERVER, config.REDIS_PORT))
```
But my scheduler seems to be firing 5 times each time, causing 5 emits. I always assumed this was due to the multiple worker processes. I even have a redis lock system set up to try and prevent this from happening, but the lock is being ignored.

**Reply by miguelgrinberg** [reply]
I suggest you ask someone familiar with your scheduler why is it firing multiple times.

**Reply by SpringsTea** [reply]
It's not the scheduler. If I run my app without wsgi, just as the flask test server, it works fine.

**Reply by miguelgrinberg** [reply]
I have no other suggestions to make. You said your scheduler was firing 5 times in your previous comment, I was just following up on your own statement.
